In [5]:
import cv2
import numpy as np

# 이미지 불러오기
img = cv2.imread('chart.jpeg')

# BGR 이미지를 HSV 이미지로 변환
hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)

# HSV 값 범위를 정의하여 빨간색을 탐지 (Hue 값은 색상환에 따라 달라질 수 있음)
lower_red = np.array([0,120,70])
upper_red = np.array([10,255,255])

# HSV 이미지에서 빨간색만 추출하기 위한 임계값 설정
mask = cv2.inRange(hsv, lower_red, upper_red)

# 비트 연산자를 사용하여 빨간색만 있는 부분을 표시
res = cv2.bitwise_and(img, img, mask=mask)

# 결과 이미지 표시
cv2.imshow('image', img)
cv2.imshow('mask', mask)
cv2.imshow('res', res)

cv2.waitKey(0)
cv2.destroyAllWindows()


## HSV 이용하는 방법

In [20]:
import cv2
import numpy as np

# 이미지 불러오기
img = cv2.imread('chart.jpeg')

# BGR 이미지를 HSV 이미지로 변환
hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)

# HSV 값 범위를 정의하여 빨간색을 탐지
lower_red = np.array([0,120,70])
upper_red = np.array([10,255,255])

# HSV 이미지에서 빨간색만 추출하기 위한 임계값 설정
mask_red = cv2.inRange(hsv, lower_red, upper_red)

# HSV 값 범위를 정의하여 파란색을 탐지
lower_blue = np.array([110,50,50])
upper_blue = np.array([130,255,255])

# HSV 이미지에서 파란색만 추출하기 위한 임계값 설정
mask_blue = cv2.inRange(hsv, lower_blue, upper_blue)

# 빨간색과 파란색 마스크 병합
mask = cv2.bitwise_or(mask_red, mask_blue)

# 비트 연산자를 사용하여 빨간색과 파란색만 있는 부분을 표시
res = cv2.bitwise_and(img, img, mask=mask)

# 결과 이미지 표시
cv2.imshow('image', img)
cv2.imshow('mask', mask)
cv2.imshow('res', res)

cv2.waitKey(0)
cv2.destroyAllWindows()


## 마우스 콜백함수

In [38]:
import cv2
import numpy as np

# 마우스 콜백 함수 정의
def get_pixel_value(event, x, y, flags, param):
    if event == cv2.EVENT_LBUTTONDOWN: # 왼쪽 버튼이 눌린 경우
        colors = img[y, x] # 해당 픽셀의 BGR 값을 가져옴
        print(f'BGR Value at ({x}, {y}): {colors}')

# 이미지 불러오기
img = cv2.imread('chart.jpeg')

# 콜백 함수 설정
cv2.namedWindow('image')
cv2.setMouseCallback('image', get_pixel_value)

while(1):
    cv2.imshow('image', img)
    if cv2.waitKey(20) & 0xFF == 27: # ESC 키를 누르면 종료
        break

cv2.destroyAllWindows()


BGR Value at (147, 1057): [113 115 180]
BGR Value at (150, 1060): [116 123 192]
BGR Value at (150, 1063): [252 250 255]
BGR Value at (153, 1067): [254 252 252]
BGR Value at (164, 1071): [142 140 200]
BGR Value at (167, 1077): [250 253 255]
BGR Value at (168, 1079): [255 254 255]
BGR Value at (172, 1083): [255 255 254]
BGR Value at (175, 1083): [251 255 250]
BGR Value at (177, 1085): [252 255 251]
BGR Value at (181, 1085): [252 248 255]
BGR Value at (184, 1084): [ 98 101 162]
BGR Value at (184, 1086): [243 242 255]
BGR Value at (186, 1089): [255 254 255]
BGR Value at (190, 1090): [255 252 255]
BGR Value at (197, 1090): [100 102 166]
BGR Value at (198, 1091): [167 168 229]
BGR Value at (200, 1093): [248 247 255]
BGR Value at (205, 1094): [208 213 255]
BGR Value at (206, 1095): [236 235 255]
BGR Value at (207, 1096): [253 245 255]
BGR Value at (207, 1097): [255 251 255]


KeyboardInterrupt: 

### img[y, x]의 순서는 img[x, y]가 아닌 img[y, x]입니다.
- BGR Value at (636, 727): [ 63  74 232] : 빨강
- BGR Value at (634, 715): [ 62  73 235]
- BGR Value at (649, 721): [218 108  50] : 파랑
- BGR Value at (691, 811): [216 106  46]
- BGR Value at (754, 807): [165  93 153] : 이평선 자주색

## RGB 이용하는 방법

In [19]:
import cv2
import numpy as np

# 이미지 불러오기
img = cv2.imread('chart.jpeg')

# RGB 값 범위를 정의하여 빨간색을 탐지
lower_red = np.array([30,40,200])
upper_red = np.array([90,100,250])

# RGB 이미지에서 빨간색만 추출하기 위한 임계값 설정
mask = cv2.inRange(img, lower_red, upper_red)

# 비트 연산자를 사용하여 빨간색만 있는 부분을 표시
res = cv2.bitwise_and(img, img, mask=mask)

# 결과 이미지 표시
cv2.imshow('image', img)
cv2.imshow('mask', mask)
cv2.imshow('res', res)

cv2.waitKey(0)
cv2.destroyAllWindows()


In [24]:
import cv2
import numpy as np

# 이미지 불러오기
img = cv2.imread('chart.jpeg')

# RGB 값 범위를 정의하여 빨간색을 탐지
lower_red = np.array([30,40,200])
upper_red = np.array([90,100,250])

# RGB 값 범위를 정의하여 파란색을 탐지
lower_blue = np.array([200, 90, 35])
upper_blue = np.array([230, 120, 60])

# RGB 이미지에서 빨간색과 파란색만 추출하기 위한 임계값 설정
mask_red = cv2.inRange(img, lower_red, upper_red)
mask_blue = cv2.inRange(img, lower_blue, upper_blue)

# 빨간색과 파란색 마스크 병합
mask = cv2.bitwise_or(mask_red, mask_blue)

# 비트 연산자를 사용하여 빨간색과 파란색만 있는 부분을 표시
res = cv2.bitwise_and(img, img, mask=mask)

# 결과 이미지 표시
cv2.imshow('image', img)
cv2.imshow('mask', mask)
cv2.imshow('res', res)

cv2.waitKey(0)
cv2.destroyAllWindows()


### 필요한 요소들을 모두 추출해보자

In [40]:
import cv2
import numpy as np

# 이미지 불러오기
img = cv2.imread('chart.jpeg')

# RGB 값 범위를 정의하여 빨간색을 탐지
lower_red = np.array([30,40,200])
upper_red = np.array([90,100,250])

# RGB 값 범위를 정의하여 파란색을 탐지
lower_blue = np.array([200, 85, 30])
upper_blue = np.array([230, 125, 70])

# RGB 값 범위를 정의하여 중심선을 탐지(초록)
lower_green = np.array([130, 60, 120])
upper_green = np.array([200, 130, 190])

# RGB 값 범위를 정의하여 하한선을 탐지
lower_yellow = np.array([140, 135, 80])
upper_yellow = np.array([190, 185, 130])

# RGB 값 범위를 정의하여 상한선을 탐지
lower_purple = np.array([70, 80, 150])
upper_purple = np.array([130, 140, 210])

# RGB 이미지에서 각 색상을 추출하기 위한 임계값 설정
mask_red = cv2.inRange(img, lower_red, upper_red)
mask_blue = cv2.inRange(img, lower_blue, upper_blue)
mask_green = cv2.inRange(img, lower_green, upper_green)
mask_yellow = cv2.inRange(img, lower_yellow, upper_yellow)
mask_purple = cv2.inRange(img, lower_purple, upper_purple)

# 모든 색상 마스크 병합
mask = cv2.bitwise_or(mask_red, mask_blue)
mask = cv2.bitwise_or(mask, mask_green)
mask = cv2.bitwise_or(mask, mask_yellow)
mask = cv2.bitwise_or(mask, mask_purple)

# 비트 연산자를 사용하여 해당 색상들만 있는 부분을 표시
res = cv2.bitwise_and(img, img, mask=mask)

# 결과 이미지 표시
cv2.imshow('image', img)
cv2.imshow('mask', mask)
cv2.imshow('res', res)

cv2.waitKey(0)
cv2.destroyAllWindows()
